### PR TITLE
Add backend model for event filters in working lists

### PR DESCRIPTION
**Pull request by ameenhere**

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/DateFilterPeriod.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import java.util.Date;
+
+/**
+ * @author Ameen Mohamed <ameen@dhis2.com>
+ *
+ */
+public class DateFilterPeriod implements Serializable
+{
+    /**
+     *
```
Comment (stian-sandvold) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/DateFilterPeriod.java]: comment?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/DateRelativePeriod.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+public enum DateRelativePeriod
+{
+    THIS_WEEK,
+    LAST_WEEK,
+    THIS_DAY,
+    LAST_DAY,
```
Comment (stian-sandvold) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/DateRelativePeriod.java]: Is it possible to reuse RelativePeriodEnum?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/ProgramStageInstanceFilter.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+/**
+ * @author Ameen Mohamed <ameen@dhis2.org>
+ *
+ */
+@JacksonXmlRootElement( localName = "programStageInstanceFilter", namespace = DxfNamespaces.DXF_2_0 )
+public class ProgramStageInstanceFilter extends BaseIdentifiableObject implements MetadataObject
+{
+
+    /**
+     *
```
Comment (stian-sandvold) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/ProgramStageInstanceFilter.java]: comment

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/ProgramStageInstanceFilterStore.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.List;
+import org.hisp.dhis.common.IdentifiableObjectStore;
+
+/**
+ * @author Ameen Mohamed <ameen@dhis2.org>
+ *
+ */
+public interface ProgramStageInstanceFilterStore extends IdentifiableObjectStore<ProgramStageInstanceFilter>
+{
+    List<ProgramStageInstanceFilter> getByProgram(String program);
```
Comment (stian-sandvold) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/ProgramStageInstanceFilterStore.java]: space between paranthesis

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+import org.hisp.dhis.security.acl.AclService;
+import org.hisp.dhis.user.CurrentUserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Ameen Mohamed <ameen@dhis2.org>
+ *
+ */
+@Transactional
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java]: Not pr. method?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+public class DefaultProgramStageInstanceFilterService implements ProgramStageInstanceFilterService
+{
+
+    // -------------------------------------------------------------------------
+    // Dependencies
+    // -------------------------------------------------------------------------
+
+    private ProgramStageInstanceFilterStore programStageInstanceFilterStore;
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java]: final + constructor?

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    public EventFilterController( ProgramStageInstanceFilterService psiFilterService, WebMessageService messageService )
+    {
+        this.psiFilterService = psiFilterService;
+        this.messageService = messageService;
+    }
+
+    /**
+     * Returns all eventFilter definitions filtered by program if provided.
+     */
+    @RequestMapping( value = "", method = RequestMethod.GET, produces = "application/json" )
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java]: You can use @GetMapping

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            throw new WebMessageException( WebMessageUtils.notFound( "EventFilter '" + uid + "' was not found." ) );
+        }
+        return psiFilter;
+    }
+
+    /**
+     * Deletes all keys with the given namespace.
+     */
+    @RequestMapping( value = "/{uid}", method = RequestMethod.DELETE )
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java]: @DeleteMapping

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            throw new WebMessageException( WebMessageUtils.notFound( "EventFilter '" + uid + "' was not found." ) );
+        }
+        psiFilterService.delete( psiFilter );
+        messageService.sendJson( WebMessageUtils.ok( "EventFilter '" + uid + "' deleted." ), response );
+    }
+
+    /**
+     * Add a new eventFilter
+     */
+    @RequestMapping( value = "", method = RequestMethod.POST, produces = "application/json", consumes = "application/json" )
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java]: @PostMapping

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    {
+        psiFilterService.add( psiFilter );
+        response.setStatus( HttpServletResponse.SC_CREATED );
+        messageService.sendJson( WebMessageUtils.created( "EventFilter created." ), response );
+    }
+
+    /**
+     * Update an eventFilter definition
+     */
+    @RequestMapping( value = "/{uid}", method = RequestMethod.PUT, produces = "application/json", consumes = "application/json" )
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java]: @PutMapping

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java (hunk 5 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java]: Class level by default. Per method only when I want it to be overridden and make it _readOnly_

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/ProgramStageInstanceFilter.java (hunk 3 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/ProgramStageInstanceFilter.java]: ok

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/DateRelativePeriod.java (hunk 2 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/DateRelativePeriod.java]: Agreed

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/DateFilterPeriod.java (hunk 1 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/DateFilterPeriod.java]: ok

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/ProgramStageInstanceFilterStore.java (hunk 4 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/programstagefilter/ProgramStageInstanceFilterStore.java]: 🤦‍♂️

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java (hunk 6 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java]: okay

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java (hunk 7 above)
Comment (ameenhere) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java]: Yea. I just used this.

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java (hunk 8 above)
Comment (ameenhere) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java]: Understood

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java (hunk 9 above)
Comment (ameenhere) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java]: Roger that

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java (hunk 10 above)
Comment (ameenhere) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventFilterController.java]: Aye Aye

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java (hunk 6 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programstagefilter/DefaultProgramStageInstanceFilterService.java]: Is this necessary? Does it give any benefits? 
Constructor with 6 arguments doesnt look good.

**Pull request merged**